### PR TITLE
Tournaments: implement banlist feature

### PR DESCRIPTION
**Pull request by sirDonovan**
Tournament creators can now specify a supplementary banlist.

Also fixes a related validator bug.

### CODE REVIEW

File: team-validator.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ class Validator {
 	constructor(format, supplementaryBanlist) {
 		format = Tools.getFormat(format);
 		if (supplementaryBanlist && supplementaryBanlist.length) {
-			format = Object.assign({}, format);
+			format = JSON.parse(JSON.stringify(format));
```
Comment (sirDonovan) [inline team-validator.js]: Regular banlists were being mutated because <code>Object.assign()</code> doesn't deep clone.

Comment (Zarel) [inline team-validator.js]: This would destroy functions in the format, though, wouldn't it?

Comment (sirDonovan) [inline team-validator.js]: Yeah, forgot about that. Updated.

File: team-validator.js [hunk 2]
```diff
@@ -22,7 +22,11 @@ class Validator {
 		if (supplementaryBanlist && supplementaryBanlist.length) {
 			format = Object.assign({}, format);
 			if (format.banlistTable) delete format.banlistTable;
-			if (!format.banlist) format.banlist = [];
+			if (format.banlist) {
+				format.banlist = format.banlist.slice(0);
```
Comment (Zarel) [inline team-validator.js]: To make a copy of an array, it's just `.slice()`.

It does the same thing as `.slice(0)` but I think it's clearer.

Comment (Zarel) [inline team-validator.js]: Although I actually feel like refactoring banlists entirely.

Comment (sirDonovan) [inline team-validator.js]: Fair enough.

What do you have in mind for the refactor?

Comment (Zarel) [inline team-validator.js]: Something like #2547 but with more modern code; probably involving `Map`s.

Comment (sirDonovan) [inline team-validator.js]: I see. Can this still be merged for now? I figure it will be easy enough to update when the refactor is fully decided.

Comment (Zarel) [inline team-validator.js]: If it's merged now, the burden of making it compatible with the refactor will fall on me... :|

...I guess that's mostly fine... Especially with how excited EG is about this.

**Pull request merged**